### ISSUE TITLE
Alpine/musl build crashes with high cpu usage due to high rounds in basic auth password

### ISSUE THREAD

**Issue by NicolasCARPi** [haproxy/haproxy#2008]
### Detailed Description of the Problem

This is a continuation of https://github.com/haproxy/haproxy/issues/1984#issuecomment-1379755788.

TL;DR: haproxy crashes 100% of the time ~2 seconds after loading the netdata service, something making quite a lot of GET requests but nothing HAproxy can't handle, especially on a server with 4 cores, and HAproxy has 4 threads.

### (docker context) The new and interesting discovery I made is that 2.7.2-alpine will crash like any other version I tried (2.4 or 2.6), but the non alpine one will NOT crash. Which is interesting.

# See edit2 below for reason of crash

### Expected Behavior

The alpine build should not crash when netdata webservice is accessed.

### Steps to Reproduce the Behavior

I could not isolate the issue enough to provide a super reproducible env/workflow/conf. It is very easy to trigger on my servers, though. So if needed I can try things there.

1. Run haproxy container, alpine flavor on Rocky Linux 9
2. Run netdata service
3. Access the netdata service => haproxy crashes immediatly

### Do you have any idea what may have caused this?

Nope but the fact that the alpine crashes and not the glibc is a good hint I guess.

### What is your configuration?

```haproxy
global
    maxconn 5000
    log stdout format raw local0 info

defaults
    mode    http

    timeout connect 5000
    timeout client  50000
    timeout server  50000

    # logs
    log global
    option  httplog
    option  dontlognull

    option forwardfor
    option http-server-close

    # use gzip compression
    compression algo gzip

# The netdata service is behind basic auth
userlist netdata-auth-list
    user netdata password XXXXXXX

# HTTP FRONTEND
frontend http-in
    bind :8080

    acl missing_cl hdr_cnt(Content-length) eq 0
    http-request deny if HTTP_URL_STAR !METH_OPTIONS || METH_POST missing_cl
    http-request deny if METH_GET HTTP_CONTENT

    # stats
    stats enable
    stats uri /stats
    stats auth YYY:ZZZZ
    stats refresh 10s
    # httpoxy.org
    http-request del-header Proxy

    # don't log HEAD requests from uptimerobot
    http-request set-log-level silent if METH_HEAD

    # evil acl
    acl evil path_beg -f /usr/local/etc/haproxy/acl-evil.lst
    # don't log the denied stuff
    http-request set-log-level silent if evil
    http-request silent-drop if evil

    # blocked ips
    acl blocked-ip src -f /usr/local/etc/haproxy/ips.d/dist/blocked.ips
    http-request set-log-level silent if blocked-ip
    http-request silent-drop if blocked-ip

    # redirect to https
    #http-request redirect scheme https code 301 unless { ssl_fc }

    # netdata
    acl netdata-acl path_beg /netdata
    # if trailing slash is missing, redirect to /netdata/
    http-request redirect scheme https drop-query append-slash if netdata-acl ! { path_beg /netdata/ }

    # followed by a few acl for services
    acl example-acl hdr(host) example.domain.com
    ...snipped...

    # Backends use
    use_backend netdata if netdata-acl


    use_backend example if example-acl

    default_backend defaultback

# BACKENDS
backend defaultback
    description Default backend

# example backend
backend example
    description example
    server example_container example_container:443 check port 443

# this is the netdata vanilla docker container on latest stable version
backend netdata
    server netdata netdata:19999 check port 19999
    acl draw-auth http_auth(netdata-auth-list)
    http-request auth realm draw unless draw-auth
    http-request set-path %[path,regsub(^/netdata/,/)]
    http-request set-header Host %[src]
    http-request set-header X-Forwarded-For %[src]
    http-request set-header X-Forwarded-Port %[dst_port]
    http-request set-header Connection "keep-alive"
```

### Output of `haproxy -vv`

```plain
This is the alpine version, the one that crashes:

 haproxy -vv

Running on: Linux 5.14.0-162.6.1.el9_1.0.1.x86_64 #1 SMP PREEMPT_DYNAMIC Mon Nov 28 18:44:09 UTC 2022 x86_64
Build options :
  TARGET  = linux-musl
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wundef -Wdeclaration-after-statement -Wfatal-errors -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wno-string-plus-int -Wno-atomic-alignment
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_GETADDRINFO=1 USE_OPENSSL=1 USE_LUA=1 USE_PROMEX=1
  DEBUG   = -DDEBUG_STRICT -DDEBUG_MEMORY_POOLS

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL +THREAD -BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H -ENGINE +GETADDRINFO +OPENSSL +LUA +ACCEPT4 -CLOSEFROM -ZLIB +SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL -SYSTEMD -OBSOLETE_LINKER +PRCTL -PROCCTL +THREAD_DUMP -EVPORTS -OT -QUIC +PROMEX -MEMORY_PROFILING

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=4).
Built with OpenSSL version : OpenSSL 3.0.7 1 Nov 2022
Running on OpenSSL version : OpenSSL 3.0.7 1 Nov 2022
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
OpenSSL providers loaded : default
Built with Lua version : Lua 5.3.6
Built with the Prometheus exporter as a service
Built with network namespace support.
Support for malloc_trim() is enabled.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.42 2022-12-11
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 12.2.1 20220924

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG

Available services : prometheus-exporter
Available filters :
        [CACHE] cache
        [COMP] compression
        [FCGI] fcgi-app
        [SPOE] spoe
        [TRACE] trace
```

### Last Outputs and Backtraces

```plain
/0 0/0 "GET /netdata/api/v1/data?chart=system.swap&_=1673492118754&format=array&points=240&group=average&gtime=0&options=absolute%7Cpercentage%7Cjsonwrap%7Cnonzero&after=-240&dimensions=used HTTP/1.1"
Thread 2 is about to kill the process.
  Thread 1 : id=0x7feb856ca9d0 act=0 glob=0 wq=1 rq=0 tl=0 tlsz=0 rqsz=0
      1/1    stuck=0 prof=0 harmless=1 wantrdv=0
             cpu_ns: poll=883220847 now=904301108 diff=21080261
             curr_task=0
* Thread 2 : id=0x7feb848a0b30 act=1 glob=1 wq=1 rq=0 tl=0 tlsz=0 rqsz=2
      1/2    stuck=0 prof=0 harmless=0 wantrdv=0
             cpu_ns: poll=9857209 now=3002220932 diff=2992363723
             curr_task=0x7feb85565cd0 (task) calls=1 last=0
               fct=0x556e6c2b02b0(process_stream) ctx=0x7feb848db770
             strm=0x7feb848db770,c08 src=192.168.1.76 fe=http-in be=netdata dst=unknown
             txn=0x7feb849012e0,3000 txn.req=MSG_BODY,4c txn.rsp=MSG_RPBEFORE,0
             rqf=64d08002 rqa=60900 rpf=a0000000 rpa=20000000
             scf=0x7feb84b388d0,EST,20 scb=0x7feb848eccc0,INI,21
             af=0,0 sab=0,0
             cof=0x7feb85556b50,80000300:H1(0x7feb855566a0)/RAW(0)/tcpv4(20)
             cob=0,0:NONE(0)/NONE(0)/NONE(-1)

  Thread 3 : id=0x7feb8487bb30 act=0 glob=0 wq=1 rq=0 tl=0 tlsz=0 rqsz=0
      1/3    stuck=0 prof=0 harmless=1 wantrdv=0
             cpu_ns: poll=2571520629 now=2589997468 diff=18476839
             curr_task=0
  Thread 4 : id=0x7feb84856b30 act=0 glob=0 wq=1 rq=0 tl=0 tlsz=0 rqsz=0
      1/4    stuck=0 prof=0 harmless=1 wantrdv=0
             cpu_ns: poll=1785718262 now=1797374678 diff=11656416
             curr_task=0
[NOTICE]   (1) : haproxy version is 2.7.1-3e4af0e
[ALERT]    (1) : Current worker (8) exited with code 134 (Aborted)
[ALERT]    (1) : exit-on-failure: killing every processes with SIGTERM
[WARNING]  (1) : All workers exited. Exiting... (134)
```

### Additional Information

Docker compose:

~~~yaml
services:
  # HAPROXY
  haproxy:
    image: haproxy:2.6.7-alpine
    container_name: haproxy
    restart: always
    logging:
      driver: syslog
      options:
        mode: non-blocking
        syslog-address: "tcp+tls://XX.YY.ZZ.WW:1514"
        tag: "haproxy-example"
        syslog-format: "rfc5424"
        syslog-tls-cert: "/etc/docker/cert.pem"
        syslog-tls-key: "/etc/docker/key.pem"
        syslog-tls-skip-verify: "true"
    read_only: true
    depends_on:
      example_service:
        condition: service_healthy
    ports:
      - '80:8080'
    volumes:
      - /YYYY/haproxy:/usr/local/etc/haproxy:ro
    networks:
      - example-net
      - external
  # END HAPROXY

# NETDATA
  netdata:
    image: netdata/netdata:v1.37.1
    hostname: example
    container_name: netdata
    restart: always
    environment:
      - NETDATA_CLAIM_TOKEN=XXXXXXXXXXXXXXXXXXX
      - NETDATA_CLAIM_URL=https://app.netdata.cloud
      - NETDATA_CLAIM_ROOMS=YYYYYYYYYYYYYYYYYY
    expose:
      - '19999'
    cap_add:
      - SYS_PTRACE
    security_opt:
      - apparmor:unconfined
    volumes:
      - /etc/os-release:/host/etc/os-release:ro
      - /proc:/host/proc:ro
      - /sys:/host/sys:ro
      - /var/run/docker.sock:/var/run/docker.sock:ro
    ulimits:
      nofile:
        soft: 1024
        hard: 1024
    networks:
      - external

networks:
  example-net:
  external:
    enable_ipv6: true
    driver: bridge
    driver_opts:
      com.docker.network.enable_ipv6: "true"
    ipam:
      driver: default
      config:
       - subnet: 2001:3984:3989::/64
~~~

**Comment by NicolasCARPi** [comment]
Good news, I am now able to reproduce this in a dev env!

The important change I had to make was to have 4 CPU. A VM with 2 CPU/threads would not crash, but 4 CPU/threads crashes. The docker-compose.yml file above is enough.

Unfortunately, I don't really manage to get something out of the core dump. The dump is generated by the host, and readable with `coredumpctl`, but of course it fails to look at the libraries that are in the container:

~~~
/usr/local/sbin/haproxy: No such file or directory.
warning: Can't open file /usr/local/sbin/haproxy during file-backed mapping note processing
warning: Can't open file /usr/lib/libpcre2-8.so.0.11.2 during file-backed mapping note processing
warning: Can't open file /usr/lib/lua5.3/liblua-5.3.so.0.0.0 during file-backed mapping note processing
warning: Can't open file /lib/libcrypto.so.3 during file-backed mapping note processing
warning: Can't open file /lib/libssl.so.3 during file-backed mapping note processing
warning: Can't open file /lib/ld-musl-x86_64.so.1 during file-backed mapping note processing
~~~

When trying to analyze the coredump directly from the container, the issue is "not in executable format: file format not recognized", because the coredump is generated by the host, so it is not compatible with stuff in the container from what I gather (I'm obviously not very knowledgeable in these areas).

If someone has an idea on how to get meaningful things from the coredump of an alpine container, please let me know!

EDIT:

So I've built haproxy without PCRE, LUA, PROMEX, GETADDRINFO or OPENSSL, installed musl-dbg and this is the best I got, by attaching gdb to the process in the container:

~~~
Attaching to process 1
Reading symbols from /usr/local/sbin/haproxy...
Reading symbols from /lib/ld-musl-x86_64.so.1...
(No debugging symbols found in /lib/ld-musl-x86_64.so.1)
0x00007f667f9f93c1 in ?? () from /lib/ld-musl-x86_64.so.1
(gdb) t a a bt full

Thread 1 (process 1 "haproxy"):
#0  0x00007f667f9f93c1 in ?? () from /lib/ld-musl-x86_64.so.1
No symbol table info available.
#1  0x00007f667f9f65ba in ?? () from /lib/ld-musl-x86_64.so.1
No symbol table info available.
#2  0x00007f667f99c31c in ?? ()
No symbol table info available.
#3  0x0000000000000008 in ?? ()
No symbol table info available.
#4  0x00007ffebc1aa080 in ?? ()
No symbol table info available.
#5  0x0000000000000003 in ?? ()
No symbol table info available.
#6  0x00007f667f79d0c0 in ?? ()
No symbol table info available.
#7  0x00007f667f9c0df8 in epoll_pwait () from /lib/ld-musl-x86_64.so.1
No symbol table info available.
#8  0x0000000000000008 in ?? ()
No symbol table info available.
#9  0x00000000135e81e6 in ?? ()
No symbol table info available.
#10 0x0000000000000000 in ?? ()
No symbol table info available.
~~~

# EDIT 2

Ok I think I found the issue, in haproxy.cfg, if I remove the basic auth:

~~~
 #acl draw-auth http_auth(netdata-auth-list)
 #http-request auth realm draw unless draw-auth
~~~

Then it doesn't crash anymore and it doesn't use a lot of CPU when loading anymore!!!

Any idea about this behavior?

**Comment by NicolasCARPi** [comment]
Ok finally pinpointed the issue, it is similar to https://github.com/haproxy/haproxy/issues/1298

The issue lies with the high number of rounds in the password hash of the auth. Using less rounds (1000) and the issue disappears!

So I'm closing this as duplicate of #1298.

EDIT: I wonder if this could be used as DOS attack, it gives an easy path to attacker to use a lot of CPU resources by sending requests to an endpoint protected with basic auth + many rounds of sha.

EDIT2: I was using password_hash from ansible which is sha512_crypt from python passlib and this has a default of [656000 rounds](https://foss.heptapod.net/python-libs/passlib/-/blob/branch/stable/passlib/handlers/sha2_crypt.py#L481).

EDIT3: I think the glibc variant doesn't crash because it is able to process more than the musl variant. I tested both with `openssl speed -evp sha512` and glibc is faster, which is no surprise because it's better optimized in some cases, like this one apparently.
